### PR TITLE
FIX (GraphEngine): @W-12138734@: Handles forEach loops executed on Set data type

### PR DESCRIPTION
**Pull request by rmohan20**
Part 2 of PR 922:
1. Creates a new interface `ApexIterableCollectionValue` to represent ApexListValue and ApexSetValue.
2. Changes to ApexForLoopValue to treat List and Set the same way.
3. New tests.

### CODE REVIEW

File: sfge/src/main/java/com/salesforce/graph/ops/TypeableUtil.java [hunk 1]
```diff
@@ -111,9 +115,23 @@ private static OrderedTreeSet getListTypeHierarchy(String definingType) {
...
+    }
+
+    private static Optional<String> getSubType(
+            Pattern subTypePattern, String definingType, int groupCountExpected) {
+        final Matcher subTypeMatcher = subTypePattern.matcher(definingType);
+
+        if (subTypeMatcher.find()) {
+            if (subTypeMatcher.groupCount() != groupCountExpected) {
+                throw new UnexpectedException(
+                        "Expected to find only one Type in declaration: " + definingType);
```
Comment (rmohan20) [inline sfge/src/main/java/com/salesforce/graph/ops/TypeableUtil.java:132]: Error message should be fixed.

File: sfge/src/main/java/com/salesforce/graph/symbols/apex/ApexForLoopValue.java [hunk 2]
```diff
@@ -31,19 +26,23 @@ public final class ApexForLoopValue extends ApexPropertiesValue<ApexForLoopValue
...
-                && !(valueVertex instanceof NewListLiteralExpressionVertex)) {
+                && !(valueVertex instanceof NewListLiteralExpressionVertex)
+                && !(valueVertex instanceof NewSetInitExpressionVertex)
+                && !(valueVertex instanceof NewSetLiteralExpressionVertex)) {
             throw new UnexpectedException(valueVertex);
         }
     }
 
-    ApexForLoopValue(ApexListValue value, ApexValueBuilder builder) {
+    ApexForLoopValue(ApexIterableCollectionValue value, ApexValueBuilder builder) {
```
Comment (rmohan20) [inline sfge/src/main/java/com/salesforce/graph/symbols/apex/ApexForLoopValue.java:36]: Most of the logic in this class remains the same. I've opened up the restraints to handle Sets the same way as Lists.

File: sfge/src/main/java/com/salesforce/graph/symbols/apex/ApexIterableCollectionValue.java [hunk 3]
```diff
@@ -0,0 +1,19 @@
+package com.salesforce.graph.symbols.apex;
+
+import com.salesforce.graph.vertex.Typeable;
+import java.util.Collection;
+import java.util.Optional;
+
+/** Represents ApexValues that hold an iterable collection of values. */
```
Comment (rmohan20) [inline sfge/src/main/java/com/salesforce/graph/symbols/apex/ApexIterableCollectionValue.java:7]: New interface to represent List and Set. Other data structures in Apex can't be iterated.

File: sfge/src/main/java/com/salesforce/graph/symbols/apex/ApexListValue.java [hunk 4]
```diff
@@ -361,19 +362,20 @@ private static Optional<Typeable> identifyType(@Nullable Typeable typeable) {
         }
 
         final String canonicalType = typeable.getCanonicalType();
-        final Matcher matcher = TYPE_PATTERN.matcher(canonicalType);
-        if (matcher.find()) {
-            if (matcher.groupCount() != 1) {
-                throw new UnexpectedException(
-                        "Expected to find only one Type in list declaration: " + typeable);
-            }
-            return Optional.of(SyntheticTypedVertex.get(matcher.group(1)));
```
Comment (rmohan20) [inline sfge/src/main/java/com/salesforce/graph/symbols/apex/ApexListValue.java:370]: Moved this logic to TypeableUtil to reuse for Sets.

File: sfge/src/main/java/com/salesforce/graph/symbols/apex/ApexSetValue.java [hunk 5]
```diff
@@ -196,4 +200,24 @@ public List<ApexValue<?>> getValues() {
     public Optional<String> getDefiningType() {
         return Optional.empty();
     }
+
+    // TODO: consolidate logic with ApexListValue
```
Comment (rmohan20) [inline sfge/src/main/java/com/salesforce/graph/symbols/apex/ApexSetValue.java:204]: There's some duplication here. Hoping to address it the next time I work in this area.

File: sfge/src/main/java/com/salesforce/graph/symbols/apex/schema/FieldSet.java [hunk 6]
```diff
@@ -141,4 +142,19 @@ private Optional<ApexValue<?>> _applyMethod(
         }
         return Optional.empty();
     }
+
+    @Override
```
Comment (rmohan20) [inline sfge/src/main/java/com/salesforce/graph/symbols/apex/schema/FieldSet.java:146]: Missing equals/hashcode were causing these items to get lost in HashSet. I've added it to all the Standard library schema types that were missing this.

File: sfge/src/test/java/com/salesforce/graph/symbols/apex/ApexSetValueTest.java [hunk 7]
```diff
@@ -360,4 +362,90 @@ void assertBoolean(
...
+                                item ->
+                                        TestUtil.apexValueToString(
+                                                ((DescribeFieldResult) item).getFieldName()))
+                        .collect(Collectors.toList());
+
+        MatcherAssert.assertThat(fieldNames, containsInAnyOrder("Name", "Phone"));
+    }
+
+    @Test
+    @Disabled // TODO: Handle method invocations on ApexClassInstanceValue
```
Comment (rmohan20) [inline sfge/src/test/java/com/salesforce/graph/symbols/apex/ApexSetValueTest.java:418]: Same problem with ApexClassInstanceValue method call as with iterating Lists.

File: sfge/src/main/java/com/salesforce/graph/symbols/apex/ApexForLoopValue.java [hunk 8]
```diff
@@ -131,7 +136,10 @@ public Optional<ApexValue<?>> apply(MethodCallExpressionVertex vertex, SymbolPro
                     valueToAdd = optApplied.get();
                 } else {
                     // TODO: path expander needs to expand on this method call and return a value
-                    valueToAdd = apexValue.deepClone(); // TODO: this is unhelpful
+                    // For now, we clone the same apexValue as a temporary bandage
+                    // to handle cases where we don't know what value a method call returns.
+                    // NOTE: The returned value is incorrect until we fix this.
```
Comment (rmohan20) [inline sfge/src/main/java/com/salesforce/graph/symbols/apex/ApexForLoopValue.java:141]: Addressing feedback in 922 about modifying the comment.

**Review by jfeingold35 — APPROVED** [review]

**Pull request merged**